typing/typedecl.ml

```diff
@@ -435,6 +432,11 @@ let transl_declaration env sdecl (id, uid) =
           Ttype_record lbls, Type_record(lbls', rep)
       | Ptype_open -> Ttype_open, Type_open
       in
+    let kind_imm = Ctype.kind_immediacy kind in
+    begin match Type_immediacy.coerce kind_imm ~as_:immediate with
+    | Ok () -> ()
+    | Error v -> raise(Error(sdecl.ptype_loc, Immediacy v))
+    end;
```

Comment:
One awkward thing about this function is that there is code to "check" the declaration error cases in three distinct places:
1. the one for the unboxed attribute above
2. the one for the immediacy attribute here
3. the one for constraints below

Case (3) is fairly different in nature, but for cases (1) and (2) it is annoying that there is no common structure to the placement: (1) is placed right when the attribute is accessed, while (2) is placed after a bunch of non-trivial logic. There is an explanation (the check (1) is self-contained from the information already there, the check (2) requires building more of the declaration to make sense), but the code remains slightly inconsistent. If I was adding support for yet another kind of attribute, where would I put the checking code? "In a random place that works" is what the current code says.

My suggestion:
- move the logic of both checks (1) and (2) to auxiliary functions, `verify_unboxed_attr` and `verify_immediacy_attr` or something
- call both those functions at the same place in the code, for example at the current point (2) or after the `let decl = ... in` declaration.

I think this would make the code both more readable and easier to extend.

Reply:
Agreed - that does sound better.  I'll take a stab at this tomorrow.

Reply:
I made the change you suggested.  However, while doing so, I realized there remains an issue.  In particular, this PR will reject the following declaration which is accepted by current ocaml.

```
type t = { x : int } [@@unboxed] [@@immediate]
```

This behavior is present in the previous iteration of this PR (#10041) but I suppose was not noticed at that time.  As the issue is that `kind_immediacy` does not traverse through `@@unboxed`, I feel this somewhat vindicates your unhappiness with that function :).

I can think of a few solutions.  The most obvious is to rename `kind_immediacy` to something like `kind_immediacy_approx`, and write a new `kind_immediacy` that does unbox things.  I think this new function would need to call `get_unboxed_type_representation` if it sees an unboxed type - unboxing just one level is potentially insufficient.  And it's possible the check needs to be moved later in `transl_type_decl` to accommodate mutually recursive or circular types (but I haven't thought that through yet).  But I thought I'd get your opinion before implementing something.

Reply:
Ah, good catch! (If you haven't done this yet, you should add the example to the testsuite.)

I agree with your suggestion to rename `kind_immediacy` into`kind_immediacy_approx` to make things more obvious. This will encourage us to consider this issue for each callsites, which we initially forgot to do ourselves.

Regarding how to fix this: my naive reaction is that we were wrong to call only `kind_immediacy` here and should have called the proper immediacy-checking function that uses the backtracking behavior, `check_decl_immediate`. Is there a reason that we cannot do this at this step, on the `decl` below? (Maybe the declaration is not fully-formed or not already in the typing environment or something, or we end up trusting the annotation before checking it?)

Reply:
OK, I've pushed a possible fix.  Here's my thinking:

We can not perform the kind check in `transl_declaration`.  The issue is mutually defined types like:
```
type t = { x : s } [@@unboxed] [@@immediate]
and s = int
```
Checking that `t`'s kind is compatible with the `[@@immediate]` attribute requires access to `s`'s manifest, but the environment used for `transl_declaration` doesn't have accurate manifests for mutually defined types.  This declaration is accepted by today's OCaml, so we don't want to reject it.

So we must perform this later - after the call to `update_type` in `transl_type_decl`.  It seems sensible to put it with the other checks in that function after the ill-formed abbrevs check.

I've done that now.  So in `transl_type_decl` we simply copy the immediacy attribute into the kind (for the abstract case), and we check it later.  Here I describe why I think the check I've added is sufficient, in combination with `check_coherence`:
- We might have an `[@@unboxed]` type, in which case we call `check_type_immediate` on the inner thing.  This inner thing might be an mutually defined abstract type whose annotation we have trusted, but in that case its manifest is checked by `check_coherence` later.
- Otherwise, we just use `kind_immediacy_approx`.  The two cases where this is innaccurate are unboxed types (already handled) and abstract types, which are either purely abstract (fine) or abbreviations (checked by `check_coherence`).

The last question is whether `check_coherence` itself can be trusted.  In particular, it checks that the manifests of abbreviations marked `[@@immediate]` are actually immediate, and the manifest may mention other abbreviation from the mutually recursive group.  But this is fine, because their manifests will also be checked.  And if there is a cycle, it's fine to treat all of them as immediate.

One consequence is that this declaration is allowed by this PR but is rejected by current OCaml:

```
type t = { x : s } [@@unboxed] [@@immediate]
and s = t [@@immediate]
```

This seems fine.  Unfortunately, this similar declaration is not allowed, and could be:

```
type t = { x : s } [@@unboxed] [@@immediate]
and s = t
```

The difference is that in the former case `s`'s kind records it is immediate, hitting the fast path of immediacy checking, while in the latter case we must expand, which runs out of fuel.  This seems acceptable, if unfortunate, to me.  But we can revisit the design if you disagree (I think it can be fixed, but at some cost in the complexity of the immediacy checking code, which doesn't seem worth it for such a silly type).